packages/gui/src/components/primitives/CalcInput.tsx

```diff
@@ -0,0 +1,187 @@
+import { kebabCase, uniq } from 'lodash-es'
+import { useEffect, useState, useId } from 'react'
+import { UnitRanges } from '../../data/ranges'
+import { convertUnits, UnitConversions, UnitSteps } from '../../lib'
+import { stringifyUnit, } from '../../lib/stringify'
+import { AbsoluteLengthUnits, CSSUnitValue, ThemeUnits, UnitlessUnits } from '../../types/css'
+import { EditorProps } from '../../types/editor'
+import { SelectInput } from '../inputs/SelectInput'
+import { Number } from '../primitives'
+import { Label } from './Label'
+import { UnitSelect } from './UnitSelect'
+import { ValueSelect } from './ValueSelect'
+
+interface CalcInputProps extends EditorProps<CSSUnitValue> {
+  label?: string
+  range?: UnitRanges
+  steps?: UnitSteps
+  units: readonly string[]
+  conversions?: UnitConversions
+  themeValues?: (CSSUnitValue & { id: string })[]
+}
+
+type CalcFunction = {
+  operand: string
+  valueX: CSSUnitValue
+  valueY: CSSUnitValue
+}
+
+export const CalcInput = ({
+  units,
+  onChange,
+  value,
+  label,
+  range,
+  steps,
+  conversions,
+  themeValues,
+}: CalcInputProps) => { 
+  const [calcOperation, setCalcOperation] = useState<CalcFunction>({
+    operand: '+',
+    valueX: value,
+    valueY: { value: 0, unit: AbsoluteLengthUnits.Px }
+  })
+  const allUnits = uniq([...units, UnitlessUnits.Number])
+
+  useEffect(() => {
+    onChange({ value: stringifyCalcValue(calcOperation), unit: UnitlessUnits.Calc })
```

Comment:
I don't think we should be returning the stringified value, right? Seems like we should be keeping the schema-based representation and then stringifying it as part of codegen?